### PR TITLE
core: rewrite agent writer on new process

### LINKED ISSUE #1102: Django API APM traces disappeared after upgrade from 0.29.0 to 0.30.0
In talking with @brettlangdon we couldn't find a quick solution, but downgrading the library version brought them back. Interestingly enough, even the custom traces we were using within the API disappeared but all the ones that exist outside of Django i.e. workers, are all fine on version 0.30.0

### PR DESCRIPTION
**Pull request by brettlangdon**
Fixes #1102 and #1103 

Previously whenever the writer noticed it was running in a new process (based on `os.getpid()`) we would reset the internal queue we used for flushing traces and we would restart the background thread we used.

We made a change to simplify this code by merging the background thread and the writer which managed it. However, when making this change we did not replicate the behavior where we recreated the background thread when we found we were running in a new process.

The changes to the architecture of `AgentWriter` are ones we wanted to keep, but there wasn't a good way to add the behavior for resetting the thread back into `AgentWriter`, since `AgentWriter` is now the background thread itself.

So this PR moves the logic of PID checking and thread resetting into `Tracer` where we already had other PID checking mechanisms anyways.

Also added a basic test case to ensure when running in a new process our tracer will recreate the background writer thread and queue.

### CODE REVIEW

**Comment by JBKahn** [comment]
Strange that the tests passed and it also exited with an error code.

**Comment by brettlangdon** [comment]
Yeah, there is something going on with our Django tests... but only in CircleCI it seems right now, can sort of ignore that failure :/

**Comment by JBKahn** [comment]
😢 well it otherwise looks pretty good

File: tests/test_tracer.py [hunk 1]
```diff
@@ -588,3 +589,36 @@ def test_tracer_dogstatsd_url():
...
+
+    def task(t, errors):
+        # Start a new span to trigger process checking
+        t.trace('test', service='test')
+
+        # Assert we recreated the writer and have a new queue
+        with capture_failures(errors):
+            assert t._pid != original_pid
+            assert t.writer != original_writer
+            assert t.writer._trace_queue != original_writer._trace_queue
```
Comment (jd) [inline tests/test_tracer.py]: can you check that we have the span we created L608 in the queue?

Comment (brettlangdon) [reply tests/test_tracer.py]: yeah, I was trying to think about how to do that, but quick enough where the background worker won't pull it from the queue before we get a chance to

I could always manually stop the background thread first... ?

File: tests/test_tracer.py [hunk 2]
```diff
@@ -588,3 +589,63 @@ def test_tracer_dogstatsd_url():
...
+    while errors.qsize() > 0:
+        raise errors.get()
+
+    # Ensure writing into the tracer in this process still works as expected
+    with t.trace('test', service='test') as span:
+        assert t._pid == original_pid
+        assert t.writer == original_writer
+        assert t.writer._trace_queue == original_writer._trace_queue
+
+        # Stop the background worker so we don't accidetnally flush the
```
Comment (majorgreys) [inline tests/test_tracer.py]:
```suggestion
        # Stop the background worker so we don't accidentally flush the
```

**Pull request merged**